[TaskManager] Add TaskManager Diceroller recipe (#13389)
microsoft/FluidFramework: docs/content/docs/recipes/task-manager-diceroller.md

    {.table}

    Run the following command to install the libraries.

    ```dotnetcli
    npm install fluid-framework @fluidframework/tinylicious-client @fluidframework/task-manager
    ```

## Code the project

1. Open the file `\src\App.tsx` in your code editor. Delete all the default `import` statements except the one that imports `App.css`. Then delete all the code in the `App()` function. The file should look like the following:

    ```ts
    import "./App.css";

    function App() {

    }

    export default App;
    ```

1. Add the following `import` statements. Note about this code:

    - `TinyliciousClient` is a Fluid service that runs on the local development computer.
    - `SharedMap` is the DDS we will use to sync the current dice roll.
    - `TaskManager` is the DDS that facilitates task distribution.

    ```ts
    import React from "react";
    import { TinyliciousClient } from "@fluidframework/tinylicious-client";
    import { ContainerSchema, IFluidContainer, ISharedMap, SharedMap } from "fluid-framework";
    import { ITaskManager, TaskManager } from "@fluid-experimental/task-manager";
    ```

### Move Fluid data to the view

> 1. The Fluid runtime will bring changes made to DDSes from any client to the current client. But Fluid is agnostic about the UI framework. You can use a helper method to get the Fluid data, from the `SharedMap` and `TaskManager` objects, into the view layer (the React state). Add the following code below the import statements. This method is called when the application loads the first time, and the value that is returned form it is assigned to a React state property.

Reviewer: ```suggestion
1. The Fluid runtime will bring changes made to DDSes from any client to the current client. But Fluid is agnostic about the UI framework. You can use a helper method to get the Fluid data, from the `SharedMap` and `TaskManager` objects, into the view layer (the React state). Add the following code below the import statements. This method is called when the application loads the first time, and the value that is returned from it is assigned to a React state property.
```

      // TODO 3: Return the Fluid SharedMap and TaskManager DDS objects.
    };
    ```

1. Replace `TODO 1` with the following code.

    ```ts
    const client = new TinyliciousClient();
    const containerSchema: ContainerSchema = {
      initialObjects: {
        sharedMap: SharedMap,
        taskManager: TaskManager,
      },
    };
    ```

1. Replace `TODO 2` with the following code. Note that `containerId` is being stored on the URL hash, and if there is no `containerId` a new container is created instead.

    ```ts
    let container: IFluidContainer;
    const containerId = window.location.hash.substring(1);
    if (!containerId) {
      ({ container } = await client.createContainer(containerSchema));
      const id = await container.attach();
      window.location.hash = id;
    } else {
      ({ container } = await client.getContainer(containerId, containerSchema));
    }
    ```

1. Replace `TODO 3` with the following code.

    ```ts
    return {
      sharedMap: container.initialObjects.sharedMap as ISharedMap,
      taskManager: container.initialObjects.taskManager as ITaskManager,
    };
    ```

### Get the Fluid data on application startup

> Now that you have defined how to get our Fluid data, you need to tell React to call `getInitialObjects()` when the application starts up and then store the result in state. Add the following at the top of the `App()` function.

Reviewer: ```suggestion
Now that you have defined how to get the Fluid data, you need to tell React to call `getInitialObjects()` when the application starts up and then store the result in state. Add the following at the top of the `App()` function.
```

  ```ts
  // TODO 1: Setup const values
  // TODO 2: Setup state
  // TODO 3: Get Fluid data on app startup
  ```

> 1. Replace `TODO 1` with the following code. These values will be used when interacting with our DDSes.

Reviewer: ```suggestion
1. Replace `TODO 1` with the following code. These values will be used when interacting with the DDSes.
```

    ```ts
    const diceKey = "dice-key";
    const diceRollTaskId = "diceRollTaskId";
    ```

2. Replace `TODO 2` with the following code. Note about this code:
   - These are the state variables that will be used by the React app.

>    - Some of these values are only used in the debug information display. Feel free to remove any that are not applicable in your applicaiton.

Reviewer: ```suggestion
   - Some of these values are only used in the debug information display. Feel free to remove any that are not applicable in your application.
```

    ```ts
    const [sharedMap, setSharedMap] = React.useState<ISharedMap>();
    const [taskManager, setTaskManager] = React.useState<ITaskManager>();
    const [diceValue, setDiceValue] = React.useState<number>();
    const [assigned, setAssigned] = React.useState<boolean>(false);
    const [queued, setQueued] = React.useState<boolean>(false);
    const [subscribed, setSubscribed] = React.useState<boolean>(false);
    ```

3. Finally, replace `TODO 3` with the following code. Note about this code:
    - Passing an empty dependency array as the last parameter of `useEffect` ensures that this function is called only once.

>     - Since `setSharedMap` and `setTaskManager` are state-changing methods, it will cause the React `App` component to immediately rerender.

Reviewer: ```suggestion
    - Since `setSharedMap` and `setTaskManager` are state-changing methods, they will cause the React `App` component to immediately rerender.
```

    ```ts
    React.useEffect(() => {
      getInitialObjects().then((initialObjects) => {
        setSharedMap(initialObjects.sharedMap);
        setTaskManager(initialObjects.taskManager);
      });
    }, []);
    ```

### Setup TaskManager logic

> Now that you have your Fluid DDSses in the React state, you can begin writing the application logic. Under the previously added code, add the following. Note about this code:

Reviewer: ```suggestion
Now that you have your Fluid DDSes in the React state, you can begin writing the application logic. Under the previously added code, add the following. Note about this code:
```

  - Passing `sharedMap` and `taskManager` in the second parameter of the `useEffect` hook ensures that the hook will not pointlessly run if the objects have not changed since the last time the `App` component rendered.

    ```ts
    React.useEffect(() => {
      if (sharedMap !== undefined && taskManager !== undefined) {
        // TODO 1: Setup state helper function
        // TODO 2: Register SharedMap handler
        // TODO 3: Setup dice roller helper function
        // TODO 4: Register TaskManager handlers
        // TODO 5: Subscribe to the task
        // TODO 6: Delete handler registrations when the React App component is dismounted.
      }
    }, [sharedMap, taskManager]);
    ```

  1. Replace `TODO 1` with the following code. This code will update the the state variables. This was consolidated into a single function for convenience.

      ```ts
      const updateState = () => {
        setDiceValue(sharedMap.get(diceKey) ?? 1);
        setAssigned(taskManager.assigned(diceRollTaskId));
        setQueued(taskManager.queued(diceRollTaskId));
        setSubscribed(taskManager.subscribed(diceRollTaskId));
      };
      ```

>   1. Replace `TODO 2` with the following code. This code will update the the state variables when the `sharedMap` object is changed.

Reviewer: Is the following what is meant?

```suggestion
  1. Replace `TODO 2` with the following code. This code will update the state variables on every client when the `sharedMap` object is changed by whatever client is currently assigned the task.
```

          <div className="debug-controls">
            <button disabled={!queued} onClick={abandon} className="debug-controls button">
              Abandon
            </button>
            <button disabled={queued} onClick={volunteer} className="debug-controls button">
              Volunteer
            </button>
            <button
              disabled={queued && subscribed}
              onClick={subscribe}
              className="debug-controls button"
            >
              Subscribe
            </button>
            <button
              disabled={!assigned}
              onClick={complete}
              className="debug-controls button"
            >
              Complete
            </button>
          </div>
        </div>
        ```

## Start the Fluid server and run the application

1. In the Command Prompt, run the following command to start the Fluid service. Note that `tinylicious` is the name of the Fluid service that runs on localhost.

    ```dotnetcli
    npx tinylicious
    ```

    If tinylicious is not installed, you will be prompted to install it. When the Fluid service is running, you will see `info: Listening on port ...` in the Command Prompt.

> 1. Open a new Command Prompt and navigate to the root of the project; for example, `C:\My Fluid Projects\collaborative-text-area-tutorial`. Start the application server with the following command. The application opens in your browser.

Reviewer: ```suggestion
1. Open a new Command Prompt and navigate to the root of the project; for example, `C:\My Fluid Projects\task-manager-diceroller-tutorial`. Start the application server with the following command. The application opens in your browser.
```

    ```dotnetcli
    npm run start
    ```

    {{< callout note >}}

    If you receive a compilation error related to a "buffer" package, then you need to install an additional dependency to make this demo compatible with Webpack 5. Run `npm install -D buffer` and try again. This will be resolved in a future release of Fluid Framework.

    {{< /callout >}}

> 1. Paste the URL of the application into the address bar of another tab or even another browser to have more than one client open at a time. Edit the text on any client and see the text change and synchronize on all the clients.

Reviewer: ```suggestion
1. Paste the URL of the application into the address bar of another tab or even another browser to have more than one client open at a time. 
```
